File: go-version/leetcode/LC1552/max_distance.go
package lc1552

import "sort"

func maxDistance(position []int, m int) int {
	sort.Ints(position)
	low, high := 1, position[len(position)-1]
	for low < high {
		mid := (low + high + 1) >> 1
		if check(position, m, mid) {
			low = mid
		} else {
			high = mid - 1
		}
	}
	return low
}

func check(position []int, m, mid int) bool {
	pre, cnt := position[0], 1
	for _, cur := range position {
		if cur-pre >= mid {
			cnt++
			pre = cur
		}
	}
	return cnt >= m
}

File: python-version/LC1552/max_distance.py
class Solution:
    def maxDistance(self, position: list[int], m: int) -> int:
        position.sort()
        low, high = 1, position[-1]
        while low < high:
            mid = (low+high+1) // 2
            if self.__check(position, m, mid):
                low = mid
            else:
                high = mid-1
        return low
    
    def __check(self, position: list[int], m : int, mid : int):
        pre, cnt = position[0], 1
        for cur in position:
            if cur - pre >= mid:
                cnt += 1
                pre = cur
        return cnt >= m